File: skills/install-slop-guard/assets/python/rules/smells/no_message_chains.py
"""Reject long chains that walk through several objects."""

from __future__ import annotations

import ast
from typing import Mapping, NamedTuple

from ...rule import Rule
from ...source import SourceFile

DEFAULT_MAX_LINKS = 3


class Chain(NamedTuple):
    """How far a chain reaches, and the name it starts from."""

    links: int
    base: str


class NoMessageChains(Rule):
    name = "no-message-chains"
    group = "smells"
    reference = "https://refactoring.guru/smells/message-chains"
    description = (
        "Disallow chains of more than `max_links` attribute or call hops, 3 by default. A chain "
        "rooted at `self` counts too, because it still walks the objects after the first hop."
    )

    def __init__(self, source: SourceFile, options: Mapping[str, object] | None = None) -> None:
        super().__init__(source, options)
        self._inner: set[int] = set()

    def visit_Attribute(self, node: ast.Attribute) -> None:
        # The walk reaches the outermost attribute first, so every hop it covers is skipped later.
        if id(node) not in self._inner:
            chain = self._measure(node)
            if chain.links > self.threshold("max_links", DEFAULT_MAX_LINKS):
                self.report(
                    node,
                    f"This walks {chain.links} objects deep from `{chain.base}`, so it depends on "
                    f"how each one is built along the way, and any of them changing breaks this "
                    f"line. Ask `{chain.base}` for what you actually need.",
                )
        self.generic_visit(node)

    def _measure(self, node: ast.expr) -> Chain:
        links = 0
        current: ast.expr = node
        while True:
            if isinstance(current, ast.Attribute):
                links += 1
                self._inner.add(id(current))
                current = current.value
            elif isinstance(current, ast.Call):
                current = current.func
            elif isinstance(current, ast.Subscript):
                current = current.value
            else:
                break
        self._inner.discard(id(node))
        base = current.id if isinstance(current, ast.Name) else self.unparse(current)
        return Chain(links, base)

File: languages/typescript/src/smells/rules/no-message-chains.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";

const DEFAULT_MAX_LINKS = 3;

/** Ban long chains that walk through several objects. */
export const noMessageChainsRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow chains of more than `maxLinks` property or call hops. A chain rooted at `this` counts too, because it still walks the objects after the first hop.",
		},
		messages: {
			messageChain:
				"This walks {{links}} objects deep from `{{base}}`, so it depends on how each one is built along the way, and any of them changing breaks this line. Ask `{{base}}` for what you actually need.",
		},
		...thresholdSchema({ maxLinks: DEFAULT_MAX_LINKS }),
	},
	createOnce(context) {
		const inner = new Set<ESTree.Node>();

		return {
			Program() {
				inner.clear();
			},
			MemberExpression(node) {
				// The walk reaches the outermost member expression first; its hops are skipped after.
				if (inner.has(node)) return;
				const limit = numberOption(context.options, "maxLinks", DEFAULT_MAX_LINKS);
				let links = 0;
				let current: ESTree.Node = node;
				while (true) {
					if (current.type === "MemberExpression") {
						links += 1;
						if (current !== node) inner.add(current);
						current = current.object;
						continue;
					}
					if (current.type === "CallExpression") {
						current = current.callee;
						continue;
					}
					if (current.type === "TSNonNullExpression" || current.type === "ChainExpression") {
						current = current.expression;
						continue;
					}
					break;
				}
				if (links <= limit) return;
				context.report({
					node,
					messageId: "messageChain",
					data: { links: String(links), base: context.sourceCode.getText(current) },
				});
			},
		};
	},
});
